kf/bpf.go

```diff
@@ -1235,3 +1204,283 @@ func ValidatePath(filePath string, destination string) (string, error) {
+			switch tmpMetricsMap.ValueSize() {
+			case 1:
+				var v int8
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 2:
+				var v int16
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 4:
+				var v int32
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 8:
+				var v int64
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			default:
+				log.Error().Msgf("unsupported map key type int16 and value size - %d", tmpMetricsMap.ValueSize())
+			}
+		} else if tmpMetricsMap.KeySize() == 4 {
+			var k int32
+			switch tmpMetricsMap.ValueSize() {
+			case 1:
+				var v int8
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 2:
+				var v int16
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 4:
+				var v int32
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 8:
+				var v int64
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			default:
+				log.Error().Msgf("unsupported map key type int32 and value size - %d", tmpMetricsMap.ValueSize())
+			}
+		} else if tmpMetricsMap.KeySize() == 8 {
+			var k int64
+			switch tmpMetricsMap.ValueSize() {
+			case 1:
+				var v int8
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 2:
+				var v int16
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 4:
+				var v int32
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 8:
+				var v int64
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			default:
+				log.Error().Msgf("unsupported key type int64 and value size - %d", tmpMetricsMap.ValueSize())
+			}
+		}
+		if err != nil {
+			return fmt.Errorf("update hash map element failed for map name %s error %v", tmpMap.Name, err)
+		}
+	}
+	return nil
+}
+
+// IsLoaded - Method verifies whether bpf program is loaded or not
+// Here it checks whether prog ID is valid and active
+func (b *BPF) IsLoaded() bool {
+	ebpfProg, err := ebpf.NewProgramFromID(b.ProgID)
+	if err != nil && errors.Is(err, os.ErrNotExist) {
+		log.Debug().Msgf("IsLoaded - %s is not loaded or invalid program id %d", b.Program.Name, uint32(b.ProgID))
+		return false
+	}
+	defer ebpfProg.Close()
+	return true
+}
+
+func (b *BPF) StartUserProgram(ifaceName, direction string, chain bool) error {
+	cmd := filepath.Join(b.FilePath, b.Program.CmdStart)
+	// Validate
+	if err := assertExecutable(cmd); err != nil {
+		return fmt.Errorf("no executable permissions on %s - error %v", b.Program.CmdStart, err)
+	}
+
+	args := make([]string, 0, len(b.Program.StartArgs)<<1)
+	args = append(args, "--iface="+ifaceName)     // attaching to interface
+	args = append(args, "--direction="+direction) // direction xdpingress or ingress or egress
+
+	if chain && b.ProgMapCollection == nil {
```

Comment:
If ProgMapCollection is defined, I'm assuming we then don't allow user programs to load and chain eBPF programs?  We probably should call this out in docs somewhere if this is the case

Reply:
We are adding in the API document in our README.